### DISCUSSION TITLE
How to get PWM running on STM32MP1 (on the Cortex-M4)

### DISCUSSION THREAD

**Discussion by JulianWeidinger** [zephyrproject-rtos/zephyr#52597] [Q&A]
Hello, 

**Problem:**
I am trying to implement a PWM LED on the Cortex M4 of the STM32MP157C-DK2 via Zephyr. Unfortunately already during the build process the following error and a number of subsequent errors come up:
...
    > In file included from /root/apps_zephyr/rtos_zephyr/soc/arm/st_stm32/common/st_stm32_dt.h:14,
    >                  from /root/apps_zephyr/rtos_zephyr/soc/arm/st_stm32/stm32mp1/soc.h:25,
    >                  from /root/apps_zephyr/rtos_zephyr/drivers/pwm/pwm_stm32.c:12:
    > /root/apps_zephyr/rtos_zephyr/drivers/pwm/pwm_stm32.c: In function 'get_tim_clk':
    > /root/apps_zephyr/build/zephyr/include/generated/devicetree_generated.h:906:37: error: 'DT_N_S_soc_S_rcc_50000000_P_apb1_prescaler' undeclared (first use in this function); did you mean 'DT_N_S_soc_S_timer_44000000_P_st_prescaler'?
    >   906 | #define DT_N_NODELABEL_rcc          DT_N_S_soc_S_rcc_50000000
    >       |                                     ^~~~~~~~~~~~~~~~~~~~~~~~~
    > /root/apps_zephyr/rtos_zephyr/include/zephyr/devicetree.h:3892:29: note: in definition of macro 'DT_CAT3'
    >  3892 | #define DT_CAT3(a1, a2, a3) a1 ## a2 ## a3
...
**Suspected cause:**
The Cortex-A7 manages as master the resources on the board and therefore the ahb prescalers for the timers are defined by the A7.
Since the Cortex-A7 runs OpenSTLinux and the Zephyr Project is built independently for the Cortex-M4, the zephyr build system does not know the prescaler. In the st,stm32mp1-rcc.yaml bindings file of the clocks the prescaler values are listed under the blocklist. So it is not possible to define the values in the device tree for the build process.

**What have I done:**

- I just configured the device tree overlay-file as followed according to the st wiki (https://wiki.st.com/stm32mpu/wiki/TIM_device_tree_configuration):
```
/{ 
         /* ... */
	soc {
		timers1: timer@44000000 {
			compatible = "st,stm32-timers";
			reg = <0x44000000 0x400>;
			clocks = <&rcc STM32_CLOCK_BUS_AHB4 0x00000002>;
			interrupts = <24 0>, <25 0>, <26 0>, <27 0>;
			interrupt-names = "brk", "up", "trgcom", "cc";
			st,prescaler = <0>;
			status = "okay";
			pwm {
				compatible = "st,stm32-pwm";
				status = "disabled";
                                pinctrl-0 = <&pwm1_pins_a>;
		                pinctrl-names = "default";
				#pwm-cells = <3>;
			};
		};
        }; 
        
	pwm1_pins_a: pwm1-0 {
	pins {
		pinmux = <STM32_PINMUX('E', 9, AF1)>;
		bias-pull-down;
		drive-push-pull;
		slew-rate = <0>;
		};
	};
	/*...*/
};

&timers1 {
	pwm1: pwm{
		status = "okay";
		pinctrl-0 = <&pwm1_pins_a>;
		pinctrl-names = "default";
	};
}; 
/*...*/
```
- To enable PWM finally, I added the following line in the prj.conf file:
`CONFIG_PWM=y`
- In the app-source code no PWM was called yet for simplification, but only the hello_world example was used.

**Question:**
How is it possible to tell the Zephyr build system at compile time the values of the prescalers?

**Comment by erwango** [comment]
@JulianWeidinger Note that the wiki note provided doesn't apply to Zephyr.

Then, unfortunately, in zephyr PWM is not yet available on STM32MP1, it should first be added in `dts/arm/st/mp1/stm32mp157.dtsi`

^^@arnopo

**Reply by JulianWeidinger** [reply]
Thanks for your answer @erwango!
I had tried the above definition of the timer1 & pwm also in the stm32mp157.dtsi directly.
I thought, it should not matter if I include the timers1 with pwm in the overlay, or in the root device tree. 
So after my attempts both works.

**Comment by arnopo** [comment]
@JulianWeidinger  

The main principle is to not use an IP on Both Linux and Zephyr

Regarding the Linux DT the timer 1 is not used by Linux
https://github.com/STMicroelectronics/linux/blob/v5.15-stm32mp/arch/arm/boot/dts/stm32mp15xx-dkx.dtsi#L628
=> So you can use it in Zephyr

Then to avoid that Linux unclock the IP you have to activate the resource manager

https://wiki.st.com/stm32mpu/wiki/Resource_manager_for_coprocessing
and enable the device in the DT

If you use the ST openSTLInux distribution you have the resource manager and timer 1 enabled for the Cortex_M4 usage
https://github.com/STMicroelectronics/linux/blob/v5.15-stm32mp/arch/arm/boot/dts/stm32mp157a-dk1-m4-examples.dts#L96

So as e first step just select this configuration with menu on boot which allows you to select the m4-exemple DT
```
Select the boot mode
1:      OpenSTLinux
2:      stm32mp157f-dk2-a7-examples
3:      stm32mp157f-dk2-m4-examples
Enter choice: 2
```
With this you should be able to manage the timer1 IP in zephyr.

Regarding the use of PWM in Zephyr, @erwango has already given you the status and method for moving forward.

**Reply by JulianWeidinger** [reply]
Thanks for your answers, too @arnopo!
You probably meant boot mode 3 "stm32mp157f-dk2-m4-examples"? 
I made it to compile the firmware now (compare post below).
Unfortunately the firmware on the board doesn't work even if I boot Linux in boot mode "stm32mp157f-dk2-m4-examples". Anyway thanks for this! I also think that this boot mode is needed!

**Comment by JulianWeidinger** [comment]
Hello @arnopo & @erwango,

Thanks for your help!
I was now able to fix the above error message and build the firmware. As mentioned in my first post, the error message occurs because the clock source for the timer comes from the RCC. In the "st,stm32mp1-rcc.yaml" it says (different from the "st,stm32-rcc.yaml") that the prescalers don't have to be defined, because the clock control comes from the A7-side. After I set the flag "compatible = "st,stm32-rcc";" at the RCC and defined the ahb, ahb1, ahb2 in the device tree, I could compile the firmware. (compare below)
... I used the default example source code for PWM (./samples/basic/blinky_pwm/) with the device tree attached below and PWM enabled in the prj.conf file.

If PWM is not implemented from Zephyr side for this board, then I won't be able to do it, since I started new with Zephyr and generally the Device Tree configuration. I now have an alternative without using the PWM module by creating a thread that simulates pwm. 
...But it is a little bit bungled.

Are there any plans to implement PWM, or the resource manager for the STM32MP157C-DK2 board?


Maybe my device tree configuration for pwm helps someone:

```
/{ 
	soc{
		timers1: timers@40000000 {
			compatible = "st,stm32-timers";
			reg = <0x40000000 0x400>;
			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00000001>;
			interrupts = <15 0>;
			interrupt-names = "global";
			st,prescaler = <0>;
			status = "okay";
		};
	};
	pwmleds {
        compatible = "pwm-leds";
        pwm_led_0: pwm_led_0 {
            pwms = <&pwm1 1 PWM_NSEC(20000000) PWM_POLARITY_INVERTED>;
            label = "PWM LED 0";
        };
    };

	aliases{
		pwm-led0 = &pwm_led_0;
    };
};

&rcc {
	compatible = "st,stm32-rcc";
	ahb-prescaler = <2>;
	apb1-prescaler = <1>;
	apb2-prescaler = <1>;
};

&timers1 {
	status = "okay";
	pwm1: pwm1{
		compatible = "st,stm32-pwm";
		status = "okay";
		#pwm-cells = <3>;
		pinctrl-0 = <&pwmLED_default>;
		pinctrl-names = "default";
	};
}; 


&pinctrl {
	pwmLED_default: pwmLED_default {
		pinmux = <STM32_PINMUX('H', 7, AF1)>;
	};
};
```

**Reply by arnopo** [reply]
> Hello @arnopo & @erwango,
> 
> Thanks for your help! I was now able to fix the above error message and build the firmware. As mentioned in my first post, the error message occurs because the clock source for the timer comes from the RCC. In the "st,stm32mp1-rcc.yaml" it says (different from the "st,stm32-rcc.yaml") that the prescalers don't have to be defined, because the clock control comes from the A7-side. After I set the flag "compatible = "st,stm32-rcc";" at the RCC and defined the ahb, ahb1, ahb2 in the device tree, I could compile the firmware. (compare below) ... I used the default example source code for PWM (./samples/basic/blinky_pwm/) with the device tree attached below and PWM enabled in the prj.conf file.
> 
> If PWM is not implemented from Zephyr side for this board, then I won't be able to do it, since I started new with Zephyr and generally the Device Tree configuration. I now have an alternative without using the PWM module by creating a thread that simulates pwm. ...But it is a little bit bungled.
> 
> Are there any plans to implement PWM, or the resource manager for the STM32MP157C-DK2 board?

The resource manager is already implemented and ready to use (no specific code should be needed on zephyr).
On Zephyr, no short term plan for the PWM.

> 
> Maybe my device tree configuration for pwm helps someone:
> 
> ```
> /{ 
> 	soc{
> 		timers1: timers@40000000 {
>```

This is the timers 2

>```
> 			compatible = "st,stm32-timers";
> 			reg = <0x40000000 0x400>;
> 			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00000001>;
> 			interrupts = <15 0>;
>```

interrupts = <28 0>;

>```
> 			interrupt-names = "global";
> 			st,prescaler = <0>;
>```

seems useless

>```
> 			status = "okay";
> 		};
> 	};
> 	pwmleds {
>         compatible = "pwm-leds";
>         pwm_led_0: pwm_led_0 {
>             pwms = <&pwm1 1 PWM_NSEC(20000000) PWM_POLARITY_INVERTED>;
>             label = "PWM LED 0";
>         };
>     };
> 
> 	aliases{
> 		pwm-led0 = &pwm_led_0;
>     };
> };
> 
> &rcc {
> 	compatible = "st,stm32-rcc";
> 	ahb-prescaler = <2>;
> 	apb1-prescaler = <1>;
> 	apb2-prescaler = <1>;
> };
> 
> &timers1 {
> 	status = "okay";
> 	pwm1: pwm1{
> 		compatible = "st,stm32-pwm";
> 		status = "okay";
> 		#pwm-cells = <3>;
> 		pinctrl-0 = <&pwmLED_default>;
> 		pinctrl-names = "default";
> 	};
> }; 
> 
> 
> &pinctrl {
> 	pwmLED_default: pwmLED_default {
> 		pinmux = <STM32_PINMUX('H', 7, AF1)>;
>```
Here is one of your issue,  the timers IP can output PWM only on some Pineouts, please refer to the datasheet
If you want to output on GPIO H7 you have to use a software PWM 

If you want to test hardware  PWM : you can use the GPIO E9 with timer 1 alreadt defined in the Linux resource manager for M4 exemple DT:
https://github.com/STMicroelectronics/linux/blob/v5.15-stm32mp/arch/arm/boot/dts/stm32mp157a-dk1-m4-examples.dts#L119

Schematic:
https://www.st.com/content/ccc/resource/technical/layouts_and_diagrams/schematic_pack/group1/36/7b/17/6e/0e/1c/4d/65/mb1272-dk2-c02_schematic/files/MB1272-DK2-C02_Schematic.PDF/jcr:content/translations/en.MB1272-DK2-C02_Schematic.PDF

some other GPIOyy/TimX ara also available on CN13 and CN14 connectors

>```
> 	};
> };
> ```

**Reply by JulianWeidinger** [reply]
Thank you for your competent comments. These were indeed very important points!

I tried your improvements and also different prescalers on the GPIO E9 with timer 1, but it didn't worked for me.
Unfortunately, I don't have time for this at the moment, because the software PWM is sufficient for me. Maybe I'll come back to that later.
